crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -80,32 +80,53 @@ impl Display for DisplayUnionType<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         let union = self.ty;
 
-        let (int_literals, other_types): (Vec<Type>, Vec<Type>) = union
+        // Group literals by type: {int, string, bytes, boolean, function, class}
+        let mut all_literals: Vec<Vec<String>> =
+            vec![vec![], vec![], vec![], vec![], vec![], vec![]];
+        let mut other_types = vec![];
+
+        union
             .elements(self.db)
             .iter()
             .copied()
-            .partition(|ty| matches!(ty, Type::IntLiteral(_)));
+            .for_each(|ty| match ty {
+                // Write a String for each literal type appended in order of appearance
+                Type::IntLiteral(n) => all_literals[0].push(format!("{n}")),
+                Type::StringLiteral(s) => all_literals[1].push(format!("\"{}\"", s.value(self.db))),
+                Type::BytesLiteral(b) => {
+                    let escape =
+                        AsciiEscape::with_preferred_quote(b.value(self.db).as_ref(), Quote::Double);
+                    if let Some(string) = escape.bytes_repr().to_string() {
+                        all_literals[2].push(string);
+                    }
+                }
+                Type::BooleanLiteral(b) => all_literals[3].push(if b {
+                    "True".to_string()
+                } else {
+                    "False".to_string()
+                }),
```

Comment:
We should actually remove the support for boolean types here, because both boolean literal types should never appear in the same union, they should be replaced by the `bool` type. See https://github.com/astral-sh/ruff/pull/13178